sound/soc/sof/intel/hda-ipc.c

```diff
@@ -254,20 +252,62 @@ irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
 	if (irq_status == 0xffffffff)
 		goto out;
 
-	/* IPC message ? */
 	if (irq_status & HDA_DSP_ADSPIS_IPC) {
 		/* disable IPC interrupt */
 		snd_sof_dsp_update_bits_unlocked(sdev, HDA_DSP_BAR,
 						 HDA_DSP_REG_ADSPIC,
 						 HDA_DSP_ADSPIC_IPC, 0);
-		ret = IRQ_WAKE_THREAD;
+		ret = true;
 	}
 
 out:
-	spin_unlock(&sdev->hw_lock);
+	spin_unlock_irq(&sdev->hw_lock);
 	return ret;
 }
 
+/* is this IRQ for ADSP or SDW? - we only care about IPC and SDW here */
+irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+	struct sof_intel_hda_dev *hdev = sdev->pdata->hw_pdata;
+	int ret = IRQ_NONE;
+
+	sdev->irq_event = 0;
+
+	/* IPC message ? */
+	if (is_ipc_irq(sdev)) {
+		/* disable IPC interrupt at is_ipc_irq() */
```

Comment:
it's quite ugly, we will use is_ipc_irq() multiple times (in the handler and in the thread), so we should only disable the interrupt once.

Reply:
That's why we have irq_event  :) Ideally, we won't call is_ipc_irq() if SOF_IRQ_IPC is set. Also, I think we need to disable irq, too if we get an interrupt on irq thread.